### PR TITLE
Add API to clear xio reference

### PR DESCRIPTION
**Pull request by danewalton**
This adds an API to clear the xio reference in the MQTT client.

Step one to fix https://github.com/Azure/azure-iot-sdk-c/issues/1666

### CODE REVIEW

File: src/mqtt_client.c [hunk 1]
```diff
@@ -945,6 +945,13 @@ static void recvCompleteCallback(void* context, CONTROL_PACKET_TYPE packet, int
     }
 }
 
+void mqtt_client_clear_xio(MQTT_CLIENT_HANDLE handle)
```
Comment (jspaith) [inline src/mqtt_client.c:948]: By convention, we need to explicitly test this against NULL and `LogError` if anything goes wrong.  

See `mqtt_client_subscribe` e.g.

Comment (jspaith) [reply src/mqtt_client.c:948]: Also we need to motivate this a bit better because someone coming to this code without context would wander why we're OK with a NULL to an allocated handle.

I'd recommend a callout along lines (pls better word smith) that "umqtt shares the xioHandle with its calling layer.  umqtt does own freeing the handle in any case and that this function is used for scenarios where caller is closing handle and wants to be positive umqtt 'gets the message'".

Again better word-smithing please :) along with whether we even refer to ourselves as umqtt in source code or just "mqtt client" or ...?

Comment (jspaith) [reply src/mqtt_client.c:948]: We might(?) even need a comment about why `mqtt_client_deinit` is not what application would call here because this brings down entire mqtt stack.  which isn't what we want on a simple network disconnect scenario.

File: inc/azure_umqtt_c/mqtt_client.h [hunk 2]
```diff
@@ -48,6 +48,7 @@ typedef void(*ON_MQTT_ERROR_CALLBACK)(MQTT_CLIENT_HANDLE handle, MQTT_CLIENT_EVE
 typedef void(*ON_MQTT_MESSAGE_RECV_CALLBACK)(MQTT_MESSAGE_HANDLE msgHandle, void* callbackCtx);
 typedef void(*ON_MQTT_DISCONNECTED_CALLBACK)(void* callbackCtx);
 
+MOCKABLE_FUNCTION(, void, mqtt_client_clear_xio, MQTT_CLIENT_HANDLE, mqtt_client);
```
Comment (jspaith) [inline inc/azure_umqtt_c/mqtt_client.h:51]: We unit test around this.  I get there's not much there but better safe than sorry.

File: src/mqtt_client.c (hunk 1 above)
Comment (danewalton) [reply src/mqtt_client.c:948]: added

File: tests/mqtt_client_ut/mqtt_client_ut.c [hunk 3]
```diff
@@ -809,6 +809,32 @@ TEST_FUNCTION(mqtt_client_deinit_succeeds)
...
+{
+    // arrange
+    MQTT_CLIENT_OPTIONS mqttOptions = { 0 };
+    SetupMqttLibOptions(&mqttOptions, TEST_CLIENT_ID, TEST_WILL_MSG, TEST_WILL_TOPIC, TEST_USERNAME, TEST_PASSWORD, TEST_KEEP_ALIVE_INTERVAL, false, true, DELIVER_AT_MOST_ONCE);
+    MQTT_CLIENT_HANDLE mqttHandle = mqtt_client_init(TestRecvCallback, TestOpCallback, NULL, TestErrorCallback, NULL);
+
+    umock_c_reset_all_calls();
+
+    // act
+    mqtt_client_clear_xio(mqttHandle);
```
Comment (jspaith) [inline tests/mqtt_client_ut/mqtt_client_ut.c:832]: Looks good.  To be extra careful could we please add another test just like this, except where there's a `mqtt_client_connect` up front (to set to a non-NULL xio, obviously a dummy ptr), then a `mqtt_client_clear_xio`, then ``mqtt_client_dowork` to make sure that the do_work happily doesn't do anything on a NULL ptr.  (And we should do the umock_c_get_expected_calls == actual_call ASSERT check to be 100% .

This reflects fact that a caller might keep happily chugging along with doWorks after a clear and that we've successfully nulled everything out.

**Review by jspaith — APPROVED** [review]

File: tests/mqtt_client_ut/mqtt_client_ut.c (hunk 3 above)
Comment (danewalton) [reply tests/mqtt_client_ut/mqtt_client_ut.c:832]: Good call. Added.
